### ISSUE TITLE
STaSy is not working on any of my datasets (I made a simple Iris colab to show the issues encountered)

### ISSUE THREAD

**Issue by AlexiaJM** [JayoungKim408/STaSy#1]
Hi,

I am trying to use STaSy, but it is not working on any of my datasets. I get a lot of values at the minimum/maximum and I often see categorical variables where one label is never generated.

To show a simple case, I made a colab file of the tutorial, but with the dataset replaced by Iris, an extremely simple dataset. Can you explain the weird behaviors and show me how to make it work on this dataset? It would be very appreciated.

I tried n_epochs=1000 and 5000, but neither work. I also tried other settings (VP, different architecture setups), but these settings did not help.

Here is the colab for Iris: https://colab.research.google.com/drive/1k5Uu1Ctc_VTWsT26uwRJ6MMydd-FFkd_?usp=sharing

**Comment by JayoungKim408** [comment]
Hi AlexiaJM,

I think the behaviors are happened due to inappropriate data pre-processing. 
I've added 2 python files for making .npz and .json file (prepare_dataset). json file created from the code contains information about each column, including min, max values, etc. The information is used in models/tabular_utils.py to pre process raw dataset. 

Please create npz and json files following the added files, and try again with the dataset. 

Thank you!

**Comment by AlexiaJM** [comment]
Hi Jayoung,

Thanks for sharing these files, but I'm actually already using the correct .npz and .json. 

The colab is using my repo here: https://github.com/AlexiaJM/STaSy_iris which is just your repo with the extra iris json and dataset. The json is here https://github.com/AlexiaJM/STaSy_iris/blob/main/tabular_datasets/iris.json and you can see that the formatting is perfect. Meanwhile the .npz is also in the correct format (numeric variables as float and categorical variables as 0, 1, 2, ... ). The only difference with yours is that the dataset is in float64 instead of float32 and my categorical variables are also float32. 

So everything should work, its indeed using the Transformer in https://github.com/JayoungKim408/STaSy/blob/main/models/tabular_utils.py to apply the min-max normalization and invert it back after sampling. You can tell that it used it because the outputs is generally either the min or max. This just means that the output the diffusion makes at the end of sampling is not in [0,1] during sampling, and must be clipped to 0 or 1, then its the normalization is inverted which leads to the min or max values. This explains the weird behaviour, but I do not know how to fix it.

Can you try to make it work using the repo https://github.com/AlexiaJM/STaSy_iris? It should be very straightforward, the data and the json is already there.

**Comment by JayoungKim408** [comment]
1. I checked the json you've made, everything is perfect but the label column. The column name should be 'label'.
2. Please refer to our paper Appendix D. Because of self-paced learning, all data samples are involved to the training after 10000 epochs. But you can change the threshold by modifying 147, 148 lines of losses.py:
`q_alpha = torch.tensor(alpha0 + torch.log( torch.tensor(1+ 0.0001718*state['step']* (1-alpha0), dtype=torch.float32) )).clamp_(max=1).to(nll.device)` and 
`q_beta = torch.tensor(beta0 + torch.log( torch.tensor(1+ 0.0001718*state['step']* (1-beta0), dtype=torch.float32) )).clamp_(max=1).to(nll.device)` to
`q_alpha = torch.tensor(alpha0 + torch.log( torch.tensor(1+ 0.01718*state['step']* (1-alpha0), dtype=torch.float32) )).clamp_(max=1).to(nll.device)` and 
`q_beta = torch.tensor(beta0 + torch.log( torch.tensor(1+ 0.01718*state['step']* (1-beta0), dtype=torch.float32) )).clamp_(max=1).to(nll.device)` (All records are used for training after 100 epochs).
Or, you can simply try Naive STaSy to get reasonable score by set config.training.spl as False.
3. But when I ran iris.json after fix the column name using your repo, I also failed to get f1 of at least 0.9 as general classifiers such as XGBoost performs on the dataset. So 1) I made my own iris.json and .npz again, 2) modified the model size (hidden_dims: (32, 64, 32), nf: 8, since the dataset has only around 200 records) and it works well on Naive STaSy (got f1 of 0.97). Your dataset and mine looks almost same, but I don't the reason yet, so I'm trying to explain it. I uploaded the files on our repo.
4. The behavior you mentioned usually happens when the model does not works well. If the model is trained well, the output of the sampling_fn is roughly within [0, 1] as the input does, so the clipping does not affect that much.

**Comment by AlexiaJM** [comment]
Thanks Jayoung! It works perfectly with those changes and also by increasing the number of epochs to 10k.
